fix: STX progress bar stuck

httpJunkie opened the pull request:
# Fix: Smart Transaction Progress Bar Not Animating

Fixes issues where smart transaction progress bars were stuck at 0% and not animating during swap submissions.

## **Description**

This PR addresses two critical issues preventing the Smart Transaction progress bar from working correctly:

### **Core Issues Fixed**

1. **useEffect Dependency Loop**: The `useRemainingTime` hook had stale closure dependencies (`isStxPastEstimatedDeadline`, `stxDeadlineSec`) that caused infinite timer restarts, preventing countdown progression.

2. **Missing creationTime for Swaps**: For swap transactions, `latestSmartTransaction` from `SmartTransactionsController` doesn't include `creationTime`, causing the timer to fail with undefined values.

### **Technical Solutions**

1. **Fixed Timer Dependencies**: 
   - Removed problematic dependencies from `useEffect` array
   - Calculate deadline values inside the timer function to avoid stale closures
   - Added regression test to prevent future timer restart issues

2. **Added creationTime Fallback**:
   - Use `useMemo` to create stable fallback timestamp when `creationTime` is missing
   - Fallback only activates when transaction becomes pending
   - Prevents component re-renders from resetting the timer

### **Testing Coverage**

Added comprehensive test suite for `useRemainingTime` hook covering:

- Basic initialization and time calculations
- Timer countdown progression over time
- Transition from estimated (45s) to max (150s) deadline
- Regression test for interval restart bug
- Edge cases for deadline transitions

## **Related issues**

Fixes:
- Smart transaction progress bar stuck at 0% during swap submission
- Progress bar animation not working on Android build 1806

## **Manual testing steps**

1. Build MetaMask Mobile with changes included
2. Perform a smart transaction swap (ETH ↔ USDC on Ethereum mainnet)
3. Verify the progress bar animates smoothly from 0% upward
4. Confirm timer countdown text updates (e.g., "Estimated completion 0:44" → "0:43")
5. Test completes successfully whether transaction finishes before or after progress bar reaches 100%

**Before**: Progress bar stuck at 0%, timer shows "0:45" and never changes
**After**: Progress bar animates smoothly, timer counts down properly

*Note: The progress bar works off estimated completion times that vary by chain. Transactions may complete before or after the progress bar reaches 100%. This is expected behavior.*

## **Screenshots/Recordings**

SEE BEFORE AND AFTER

### **Before**

https://github.com/user-attachments/assets/4105f633-a612-49b6-b408-0e570df1f9d7

### **After**

https://github.com/user-attachments/assets/46841068-c7e0-461a-ab4f-716bea61f20f

(Progress Bar May not always make it to 100%. This is ok as time remaining is estimate)

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

httpJunkie committed:
fix(swaps): Fix STX progress bar stuck due to useEffect dependencies

Remove stale closure dependencies causing interval restarts in
useRemainingTime hook. Progress bar now properly animates during
Smart Transaction submission.

Fixes issue where progress bar was no animating at all.

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing

httpJunkie committed:
fix(swaps): Add ESLint disable for STX timer dependencies

Disable react-hooks/exhaustive-deps warning in useRemainingTime.
Adding isStxPastEstimatedDeadline and deadline dependencies would
cause useEffect to restart the timer interval whenever state changes,
preventing countdown and breaking progress bar animation.

Timer function reads fresh state values to avoid stale closures
while keeping stable dependencies to prevent restart loops.

httpJunkie committed:
Merge branch 'fix/progress-bar-not-progressing' of github.com:MetaMask/metamask-mobile into fix/progress-bar-not-progressing

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing

httpJunkie committed:
test: add unit tests for useRemainingTime hook

httpJunkie committed:
test: fix linting issues in useRemainingTime tests

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2

httpJunkie committed:
fix(stx): Add creationTime fallback for swap progress bar animation

Swap transactions may not include creationTime in latestSmartTransaction,
causing progress bar to stay at 0%. Add useMemo fallback using current
timestamp when transaction becomes pending to ensure stable timer operation.

Fixes progress bar animation for Smart Transaction swaps.

httpJunkie committed:
Merge branch 'fix/progress-bar-not-progressing-2' of github.com:MetaMask/metamask-mobile into fix/progress-bar-not-progressing-2

httpJunkie committed:
fix(smart-tx): correct remaining time calculation and add robust deadline transition tests

- Fix useRemainingTime to properly handle transition from estimated to max deadline, ensuring no negative countdowns and clamping to zero when past max deadline.
- Add/expand tests to cover deadline transitions, countdown, and edge cases.
- All tests now pass and logic matches expected smart transaction progress bar behavior.

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2

httpJunkie committed:
fix(stx): Stabilize effectiveCreationTime fallback across status changes

Remove isStxPending dependency to prevent timer resets if status
flickers. Addresses CursorBot review feedback.

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2

dan437 commented on app/components/Views/transactions/SmartTransactionStatus/SmartTransactionStatus.tsx:
```diff
@@ -1,4 +1,4 @@
-import React from 'react';
+import React, { useMemo} from 'react';
```
nit: missing a space after useMemo

dan437 commented on app/components/Views/transactions/SmartTransactionStatus/useRemainingTime.test.tsx:
```diff
@@ -0,0 +1,186 @@
+  selectSwapsChainFeatureFlags: jest.fn(() => undefined),
+}));
+
+// Create a wrapper for tests
```
Why is this comment needed? `const createWrapper` already says pretty much the same thing.

httpJunkie committed:
fix: pr issues

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2

httpJunkie committed:
fix(stx): Prevent timer memory leak when past max deadline

Replace stale closure check with fresh elapsed time calculation
to ensure interval clears properly after 150 seconds. Update
test to verify boundary condition behavior.

Fixes indefinite timer when transactions exceed max deadline.

httpJunkie committed:
fix: Only provide fallback if transaction is pending, otherwise returns an `undefined`. This prevents associating the timer to the component mount

httpJunkie committed:
Merge branch 'main' into fix/progress-bar-not-progressing-2
